### ISSUE TITLE
Consider using git subtree

### ISSUE THREAD

**Issue by jcbhmr** [spenserblack/actions-wiki#11]
```sh
remote="$GITHUB_SERVER_URL/$GITHUB_REPOSITORY.wiki.git" # Or similar string concatenation
ref=$(git subtree split -P "$INPUT_PATH")
git push -f "$remote" "$ref:master"
```

This would...

1. Keep a mirrored history of all the commits with proper authors and proper history for free since you are just "splitting out" a specific subfolder of git
2. Make it so you have to commit things to the repo if they are generated by bots in the action itself (like a generated CHANGELOG.md that isn't committed, but want to post it to the wiki)
3. Avoid the need to take inputs about the author and commit message and everything

Note that 2 is likely a downside, but could be mitigated; you can either force the user to use a commit action or git add CHANGELOG.md && git commit -am "Create changelog" themselves, or automagically run git commit in the action itself.

This probably would warrant a version bump. Currently, you're at v0.

![image](https://user-images.githubusercontent.com/61068799/224774674-c4e7f098-ec57-43b1-8f49-758eb5061441.png)

I'd love to contribute code to do this. ❤

/ref https://github.com/Andrew-Chen-Wang/github-wiki-action/issues/18

**Comment by jcbhmr** [comment]
When/if this happens, add "🌲 Uses `git subtree` to preserve history even in the wiki viewer" to the readme emoji bullet point list thing

**Comment by spenserblack** [comment]
Some thoughts:
In https://github.com/spenserblack/actions-wiki/issues/8#issue-1619587195 you posted an example where git subtree is used. As far as I could tell, the wiki's commits did have the same messages and authors, but *different SHAs.* Using the message(s) of the workflow's trigger is cool, but I think it's a reasonable default to include the SHAs in the wiki commit messages. This would act like a sort of "proof" where users can link their wiki history to their repo history, if necessary. Just something to consider, but I don't think it should block subtree-ing if it's too much work.

>Avoid the need to take inputs about the author and commit message

I think that any action that makes commits should give users the ability to set the commit message via an input. In fact, this is making me think that not having `author-name` and `author-email` inputs was an oversight that I should add in :laughing:

Right now this action is basically "squashing" changes in the wiki. This is personally what I like, but I can definitely see how other users would prefer to "subtree". So I'm thinking that we can have an input called `strategy` (I'm open to suggestions on the name :slightly_smiling_face:), with possible values `squash` and `subtree`. Even though this is v0, I'd like to avoid a breaking change, so keep the default value at `squash`. The value of `inputs.strategy` can dynamically pick which script to run. Maybe even `./${{ inputs.strategy }}.sh`. We better not ever have `./secret-script-that-exposes-our-secrets.sh` if we do that :laughing: Probably best to have some sort of validation anyway.

Perhaps the `commit-message` input would be valid only for the "squash" strategy.

**Comment by jcbhmr** [comment]
> In [#8 (comment)](https://github.com/spenserblack/actions-wiki/issues/8#issue-1619587195) you posted an example where git subtree is used. As far as I could tell, the wiki's commits did have the same messages and authors, but _different SHAs_

This is **correct** because it literally reconstructs a new wholly original orphaned branch. At least, that's my understanding...

>  I think it's a reasonable default to include the SHAs in the wiki commit messages. This would act like a sort of "proof" where users can link their wiki history to their repo history, if necessary.

True.... but honestly I don't think this happens all the much. It's a good valid concern, and certainly a _caveat_, but I think that the commit messages **and dates** (dates are probably _more_ important, especially for nontechies who just want a rough idea of who and when) are enough for most cases. You can also put in the "alternatives" section that "there are other actions where you can generate a new commit message and customize the committer to a greater extent" or similar?

If I'm being totally honest, the reason I used git subtree over the traditional copy-to-a-folder-then-git-init-the-remote-add-then-add-all-then-message-then-commit is because _it was simply so much easier to do_ in three lines vs like 40 to get all the processing of constructing, copying, validating, pushing, authenticating, etc.

> I think that any action that makes commits should give users the ability to set the commit message via an input. In fact, this is making me think that not having author-name and author-email inputs was an oversight that I should add in 😆

yeah you can check out the wealth of options and complexity that commits can have in https://github.com/EndBug/add-and-commit which might be a good case to use and just pass-through all the options that the user gives straight into that action. Since we currently use a composite action, we can do that!

> Right now this action is basically "squashing" changes in the wiki. This is personally what I like, but I can definitely see how other users would prefer to "subtree". So I'm thinking that we can have an input called strategy (I'm open to suggestions on the name 🙂), with possible values squash and subtree. Even though this is v0, I'd like to avoid a breaking change, so keep the default value at squash. The value of inputs.strategy can dynamically pick which script to run. Maybe even ./${{ inputs.strategy }}.sh. We better not ever have ./secret-script-that-exposes-our-secrets.sh if we do that 😆 Probably best to have some sort of validation anyway.
>
> Perhaps the commit-message input would be valid only for the "squash" strategy.

Sounds like a good name for an option. I'd first look at all the official actions and maybe @wow-actions to see if there's any prior art of names (consistency is often -- but not always -- good)

This sounds remarkably similar to Andrew-Chen-Wang's options that are being discussed.... 🤔

**Comment by jcbhmr** [comment]
> Even though this is v0, I'd like to avoid a breaking change,

I don't think this should concern you in the slightest. You have literally 4 users. Three of which are yourself. 😉

**opinion time** And it's _OK to make breaking changes_ you're not locked in to the past. This is something you would def need to consider if you had automatic rollover (like from v1.1.1 to v1.1.2 when using a v1 tag) but I think now is the perfect time to bump major versions, play with settings, and see what ideas stick. In this case, that means playing around with git stuff and PRs for a few days/weeks before releasing v1 and some nice release notes. But that's how I'd do it. I'm not you.

**Comment by spenserblack** [comment]
>You have literally 4 users. Three of which are yourself.

75% of my userbase wants `squash` to be the default strategy :rofl:

But on a serious note, regardless of whether or not it's breaking, I'm leaning towards squash being the default because it's closer to the behavior I had in mind. That behavior was "Just update the wiki, my history is in the repo" :laughing: I'm obviously biased since it's the default *I* want. But I think it might be a bit easier to migrate to using this action using the "squash strategy" (like what happened on onefetch). Plus the squash strategy allows one to control the commit message, since it's guaranteed to be one commit per push.

I should note that I'm probably influenced by [deploying to GitHub Pages](https://cli.vuejs.org/guide/deployment.html#github-pages), where the src history mattered, but the destination history didn't matter nearly as much.

**Comment by jcbhmr** [comment]
> 75% of my userbase wants squash to be the default strategy 🤣

I'll be honest, the squash strategy is growing on me... I just **heavily** dislike the idea of a commit message. Heck, why is there even a commit message to set? You hardly _ever_ see it...

Radical proposal (I might regret or backtrack this later): use the `env` field as kinda like _super-mega-secret-options_ sorta like how Git uses them to configure commits. Basically I'm saying that we can re-use the kinda infra the Git has laid out. You **rarely** need to modify it, and you **rarely** should need to, but sometimes you do.

https://git-scm.com/book/en/v2/Git-Internals-Environment-Variables#:~:text=case%2Dinsensitive%20manner.-,Committing,-The%20final%20creation

**Comment by jcbhmr** [comment]
Since this is as good a place as any, here's the .zip of my jcbhmr/publish-wiki-action repo that I have since deleted. 
[publish-wiki-action-OLD-main.zip](https://github.com/spenserblack/actions-wiki/files/10963384/publish-wiki-action-OLD-main.zip)

**Comment by jcbhmr** [comment]
> > Heck, why is there even a commit message to set?
> 
> Well, this action has a default commit, so the user doesn't _need_ to set it.

My bad, I was conflating this repo with Andrew-Chen-Wang/github-wiki-action where you are _required_ to set token, name, commit message, and author email. Sorry! 😬 I directed my frustration at the wrong source.

Yes, I do think that your default commit message is pretty good. SHA in message + date from metadata seems good 👍

> I think I wouldn't want to make the user use env:, because it's less discoverable IMO than inputs. It would also make the user read some docs to know what environment variables are OK, while they can just read the inputs: field in action.yml to know what inputs are supported.

That was kinda the goal of the proposal lmao: make it _dilberately "secret"_ to discourage use kinda like how git does it for setting/overriding commit author stuff.

But yeah I'm kinda indifferent now to my original radical idea. Coming back after a day really can switch your perspective. If you want to put them in inputs, fine by me. I just kinda figured that there was an existing convention to override that sort of stuff. Particularly the difference between things like commit message and commit author. I could see "message" being an input field, with all the Git author stuff being env. Or all inputs like https://github.com/EndBug/add-and-commit

**Comment by jcbhmr** [comment]
To bring this back on-topic tho back to using git subtree vs copy+commit

I think that the idea of git subtree _is interesting_ and _could be an option_ like you mentioned "strategy", but I think I'm warming to your comparison of `actions/deploy-pages` where it just grabs a folder and pushes it to a website; similarly this action should just grab a folder and push to a wiki.

Actually, this brings back my naming idea lmao #14 

---

but on a serious note, I think this issue comes down to the question

#### Do you want to implement a `strategy` option?

if yes = pr and close issue
if no = close as unplanned

As you may have noticed, I'm interested in assisting. 😜
